emlid/devlib
# Allow users without admin privileges open storage device on macOS

**GoodDay:** [Flasher for Mac Can't Connect to M2](https://www.goodday.work/t/ehulLE)

## Motivation
Non-admin users are unable to reflash devices, because `sudo --askpass` doesn't work and they can't open storage devices for writing.
In general, there are 2 problems. Users without admin privileges:
1) Unable to unmount disk mount points.
2) Unable to open storage devices

## Solution
For the "1)" problem solution is [DADiskUnmount](https://developer.apple.com/documentation/diskarbitration/1492758-dadiskunmount?language=objc) function, which allows to unmout the whole disk. It's enough to be able to write.
An example was picked from https://github.com/raspberrypi/rpi-imager/blob/8a44c6b3ca99b87896451d7fe48342955675bad4/dependencies/mountutils/src/darwin/functions.cpp#L171


For the "2)" problem solution is `/usr/libexec/authopen` util, which allows entering admin password and get a special fd for storage file.
An example was picked from https://github.com/raspberrypi/rpi-imager/blob/8a44c6b3ca99b87896451d7fe48342955675bad4/mac/macfile.cpp#L25

## How to test
1) Get a storage device, that has mount points. For example, I picked ReachRS2.
2) Put the device into storage mode
3) Patch `devlib_cli` example. Here is the patch: 
[changed_devlib_cli.txt](https://github.com/emlid/devlib/files/5844038/changed_devlib_cli.txt)

> To apply the patch use the following command: `$ git apply changed_devlib_cli.txt`
4) Run `devlib_cli` and check the result. You should see something like:
```shell
$  ./devlib_cli
...output...
-> Use device on "20-3.4" port
...output...
-> macx_native: Unmounting
-> macx_native: Schedule session on run loop
-> macx_native: Waiting run loop
-> macx_native: Unmount callback
-> macx_native: Unmount success
...output...
-> Opening: false
...output...
-> Auth Opening: true
``` 

## Changes
- `devlib/native`:
    - `umount` -> `umountParition`
    - `umountDisk` function has been added and implemented on macOS
    - `authOpen` function has been added and implemented on macOS
- `StorageDeviceFile`:
    - `authOpen` function has been added
    - `StorageDeviceFileImpl` tries unmount the whole disk instead of mountpoints
    - unused 'mntptLock' object member has been removed from `StorageDeviceFileImpl`
- `devlib_cli`:
    - `QCoreApplication` events loop has been removed
